From 54fe98a29c74aac9cf7daed7f70e230b7bbcb955 Mon Sep 17 00:00:00 2001
From: J. Bruce Fields <bfields@redhat.com>
Date: Tue, 16 Sep 2014 20:20:24 -0400
Subject: [fs] nfsd: don't destroy clients that are busy

Message-id: <1410898849-22126-141-git-send-email-bfields@redhat.com>
Patchwork-id: 95750
O-Subject: [Patch RHEL7 140/165] nfsd: don't destroy clients that are busy
Bugzilla: 1044798 1123721 1141932
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Jeff Layton <jlayton@primarydata.com>

It's possible that we'll have an in-progress call on some of the clients
while a rogue EXCHANGE_ID or DESTROY_CLIENTID call comes in. Be sure to
try and mark the client expired first, so that the refcount is
respected.

This will only be a problem once the client_mutex is removed.

Signed-off-by: Jeff Layton <jlayton@primarydata.com>
Signed-off-by: J. Bruce Fields <bfields@redhat.com>
Upstream: fd699b8a48c0ca36e782cf705794358b3e4b8c25
Bugzilla: 1044798
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfsd/nfs4state.c b/fs/nfsd/nfs4state.c
index 028ae55..037bb92 100644
--- a/fs/nfsd/nfs4state.c
+++ b/fs/nfsd/nfs4state.c
@@ -2267,8 +2267,11 @@ nfsd4_exchange_id(struct svc_rqst *rqstp,
 
  /* case 1 (normal case) */
 out_new:
- if (conf)
-  unhash_client_locked(conf);
+ if (conf) {
+  status = mark_client_expired_locked(conf);
+  if (status)
+   goto out;
+ }
  new->cl_minorversion = cstate->minorversion;
  new->cl_mach_cred = (exid->spa_how == SP4_MACH_CRED);
 
@@ -2881,6 +2884,9 @@ nfsd4_destroy_clientid(struct svc_rqst *rqstp, struct nfsd4_compound_state *csta
    status = nfserr_clientid_busy;
    goto out;
   }
+  status = mark_client_expired_locked(conf);
+  if (status)
+   goto out;
   clp = conf;
  } else if (unconf)
   clp = unconf;
-- 
1.7.1